src/Main.elm

```diff
@@ -628,21 +653,108 @@ raisedDiv =
         ]
 
 
-modalView : Model -> Html Msg
-modalView model =
+type Menu
+    = DifficultyMenu Difficulty GridProperties
+
+
+type MenuMsg
+    = SetDifficulty Difficulty
+
+
+updateMenu : MenuMsg -> Menu -> Menu
+updateMenu msg (DifficultyMenu difficulty fields) =
+    case msg of
+        SetDifficulty (Custom properties) ->
+            DifficultyMenu (Custom properties) properties
+
+        SetDifficulty d ->
+            DifficultyMenu d fields
+
+
+viewModal : Model -> Menu -> Html Msg
+viewModal model (DifficultyMenu difficulty fields) =
+    let
+        menuContent =
+            Html.map MenuMsg <|
+                div []
+                    [ radiobutton "Beginner" (difficulty == Beginner) Beginner
+                    , radiobutton "Intermediate" (difficulty == Intermediate) Intermediate
+                    , radiobutton "Expert" (difficulty == Expert) Expert
+                    , radiobutton "Custom" (isCustom difficulty) (Custom fields)
+                    , viewCustomFields (isCustom difficulty) fields
+                    ]
+    in
     div maskStyle
         [ modalContent []
-            [ windowsChrome [ style "padding" "0 18px 18px" ]
+            [ windowsChrome [ style "padding" "0 18px 90px" ]
                 [ formGroup "Difficulty"
-                    [ radiobutton "Beginner" Beginner model.game
-                    , radiobutton "Intermediate" Intermediate model.game
-                    , radiobutton "Expert" Expert model.game
+                    [ menuContent
+                    , button [ onClick (CloseMenu (Just difficulty)) ] [ text "OK" ]
+                    , button [ onClick (CloseMenu Nothing) ] [ text "Cancel" ]
                     ]
                 ]
             ]
         ]
 
 
+radiobutton : String -> Bool -> Difficulty -> Html MenuMsg
+radiobutton settingLabel isSelected difficulty =
+    label [ style "display" "flex", style "align-items" "center" ]
+        [ input
+            [ type_ "radio"
+            , name "value"
+            , onClick (SetDifficulty difficulty)
+            , checked isSelected
+            , style "margin" "4px 8px"
+            ]
+            []
+        , text settingLabel
+        ]
+
+
+viewCustomFields : Bool -> GridProperties -> Html MenuMsg
+viewCustomFields enabled ({ width, height, bombs } as fields) =
+    let
+        toInput num onInput_ =
+            input
+                [ type_ "number"
+                , value (String.fromInt num)
+                , onInput
+                    (String.toInt
+                        >> Maybe.withDefault 0
+                        >> onInput_
```

Comment:
...which means i'm more confident here that i'll be able to parse the string that comes out. using `Maybe.withDefault 0` just means "if you can't parse it, give me a zero". 

now that the uncertainty is out of the way, `onInput_` can stay simple since it's definitely receiving an `Int`